mmdet3d/models/dense_heads/groupfree3d_head.py

```diff
@@ -0,0 +1,1000 @@
+                    num_reg_out_channels=self._get_reg_out_channels()))
+
+        self.sampling_objectness_loss = build_loss(sampling_objectness_loss)
+        self.objectness_loss = build_loss(objectness_loss)
+        self.center_loss = build_loss(center_loss)
+        self.dir_res_loss = build_loss(dir_res_loss)
+        self.dir_class_loss = build_loss(dir_class_loss)
+        self.semantic_loss = build_loss(semantic_loss)
+        if self.size_cls_agnostic:
+            self.size_reg_loss = build_loss(size_reg_loss)
+        else:
+            self.size_res_loss = build_loss(size_res_loss)
+            self.size_class_loss = build_loss(size_class_loss)
+
+    def init_weights(self):
+        """Initialize weights of transformer decoder in GroupFree3DHead."""
+        # initialize transformer
+        for m in self.decoder_layers.parameters():
+            if m.dim() > 1:
+                nn.init.xavier_uniform_(m)
+
+        for m in self.decoder_self_posembeds.parameters():
+            if m.dim() > 1:
+                nn.init.xavier_uniform_(m)
+
+        for m in self.decoder_cross_posembeds.parameters():
+            if m.dim() > 1:
+                nn.init.xavier_uniform_(m)
+
+    def _get_cls_out_channels(self):
+        """Return the channel number of classification outputs."""
+        # Class numbers (k) + objectness (1)
+        return self.num_classes + 1
+
+    def _get_reg_out_channels(self):
+        """Return the channel number of regression outputs."""
+        # center residual (3),
+        # heading class+residual (num_dir_bins*2),
+        # size class+residual(num_sizes*4 or 3)
+        if self.size_cls_agnostic:
+            return 6 + self.num_dir_bins * 2
+        else:
+            return 3 + self.num_dir_bins * 2 + self.num_sizes * 4
+
+    def _extract_input(self, feat_dict):
+        """Extract inputs from features dictionary.
+
+        Args:
+            feat_dict (dict): Feature dict from backbone.
+
+        Returns:
+            torch.Tensor: Coordinates of input points.
+            torch.Tensor: Features of input points.
+            torch.Tensor: Indices of input points.
+        """
+
+        seed_points = feat_dict['fp_xyz'][-1]
+        seed_features = feat_dict['fp_features'][-1]
+        seed_indices = feat_dict['fp_indices'][-1]
+
+        return seed_points, seed_features, seed_indices
+
+    def forward(self, feat_dict, sample_mod):
+        """Forward pass.
+
+        Note:
+            The forward of GroupFree3DHead is devided into 2 steps:
+
+                1. Initial object candidates sampling.
+                2. Iterative object box prediction by transformer decoder.
+
+        Args:
+            feat_dict (dict): Feature dict from backbone.
+            sample_mod (str): sample mode for initial candidates sampling.
+
+        Returns:
+            results (dict): Predictions of GroupFree3D head.
+        """
+        assert sample_mod in ['fps', 'kps']
```

Comment:
Should sample_mod be defined during init?

Reply:
VoteNet puts sample_mod in forward probably because train-time and test-time are different. Then I think this is reasonable.

Reply:
Wondering in GroupFree3D the sample_mod could also be different during train and test? although it's not the normal case